tests/approaches/test_grammar_search_invention_approach.py

```diff
@@ -147,11 +150,275 @@ def test_unary_free_forall_classifier():
+    assert handempty_included_h > none_included_h # this is very bad!
+
+
+def test_hadd_lookahead_heuristic():
+    """Tests for _HaddLookaheadHeuristic().
+    """
+    # Tests for CoverEnv.
+    utils.update_config({
+        "env": "cover",
+        "offline_data_method": "demo+replay",
+        "seed": 0,
+    })
+    env = CoverEnv()
+    ablated = {"HandEmpty", "Holding"}
+    initial_predicates = set()
+    name_to_pred = {}
+    for p in env.predicates:
+        if p.name in ablated:
+            name_to_pred[p.name] = p
+        else:
+            initial_predicates.add(p)
+    candidates = {p: 1.0 for p in name_to_pred.values()}
+    dataset = create_dataset(env, next(env.train_tasks_generator()))
+    atom_dataset = utils.create_ground_atom_dataset(dataset, env.predicates)
+    heuristic = _HaddLookaheadHeuristic(initial_predicates, atom_dataset,
+                                        candidates)
+    all_included_h = heuristic.evaluate(set(candidates))
+    handempty_included_h = heuristic.evaluate({name_to_pred["HandEmpty"]})
+    holding_included_h = heuristic.evaluate({name_to_pred["Holding"]})
+    none_included_h = heuristic.evaluate(set())
+    assert all_included_h < holding_included_h < none_included_h
+    assert all_included_h < handempty_included_h  # not better than none
+
+    # Test that the score is inf when the operators make the data impossible.
+    ablated = {"Covers"}
+    initial_predicates = set()
+    name_to_pred = {}
+    for p in env.predicates:
+        if p.name in ablated:
+            name_to_pred[p.name] = p
+        else:
+            initial_predicates.add(p)
+    candidates = {p: 1.0 for p in name_to_pred.values()}
+    # Reuse dataset from above.
+    heuristic = _HaddLookaheadHeuristic(initial_predicates, atom_dataset,
+                                        candidates)
+    assert heuristic.evaluate(set()) == float("inf")
+
+    # Tests for BlocksEnv.
+    utils.flush_cache()
+    utils.update_config({
+        "env": "blocks",
+        "offline_data_method": "demo+replay",
+        "seed": 0,
+    })
+    env = BlocksEnv()
+    ablated = {"Holding", "Clear", "GripperOpen"}
+    initial_predicates = set()
+    name_to_pred = {}
+    for p in env.predicates:
+        if p.name in ablated:
+            name_to_pred[p.name] = p
+        else:
+            initial_predicates.add(p)
+    candidates = {p: 1.0 for p in name_to_pred.values()}
+    dataset = create_dataset(env, next(env.train_tasks_generator()))
+    atom_dataset = utils.create_ground_atom_dataset(dataset, env.predicates)
+    heuristic = _HaddLookaheadHeuristic(initial_predicates, atom_dataset,
+                                        candidates)
+    all_included_h = heuristic.evaluate(set(candidates))
+    none_included_h = heuristic.evaluate(set())
+    # Note: the values for Holding alone, Clear alone, and GripperOpen alone
+    # are not in between the bounds. Here are all the values:
+    # ipdb> all_included_h
```

Comment:
Did you use your DebugGrammar stuff to quickly figure out these values?

Reply:
yeah